internal/monitors/metricsfilter.go

```diff
@@ -0,0 +1,153 @@
+		m, err := filter.NewBasicStringFilter([]string{metricName})
+		if err != nil {
+			return err
+		}
+
+		for metric := range metadata.Metrics {
+			if m.Matches(metric) {
+				return nil
+			}
+		}
+
+		return fmt.Errorf("metric pattern '%s' did not match any available metrics", metricName)
+	}
+
+	if !metadata.HasMetric(metricName) {
+		return fmt.Errorf("metric '%s' does not exist", metricName)
+	}
+
+	return nil
+}
+
+func validateGroup(metadata *Metadata, group string) ([]string, error) {
+	if strings.TrimSpace(group) == "" {
+		return nil, errors.New("group cannot be empty")
+	}
+
+	metrics, ok := metadata.GroupMetricsMap[group]
+	if !ok {
+		return nil, fmt.Errorf("group %s does not exist", group)
+	}
+	return metrics, nil
+}
+
+func newMetricsFilter(metadata *Metadata, extraMetrics, extraGroups []string) (*metricsFilter, error) {
+	var filterItems []string
+
+	for _, metric := range extraMetrics {
+		if err := validateMetricName(metadata, metric); err != nil {
+			return nil, err
+		}
+
+		// If the user specified a metric that's already included no need to add it.
+		if !metadata.IncludedMetrics[metric] {
+			filterItems = append(filterItems, metric)
+		}
+	}
+
+	for _, group := range extraGroups {
+		metrics, err := validateGroup(metadata, group)
+		if err != nil {
+			return nil, err
+		}
+		filterItems = append(filterItems, metrics...)
+	}
+
+	basicFilter, err := filter.NewBasicStringFilter(filterItems)
+	if err != nil {
+		return nil, fmt.Errorf("unable to construct filter with items %s: %s", filterItems, err)
+	}
+
+	effectiveMetrics := map[string]bool{}
+
+	// Precompute set of metrics that matches the filter. This isn't a complete
+	// set of metrics that are enabled in the case of metrics that aren't included
+	// in metadata. But it provides a fast path for known metrics.
+	for metric := range metadata.Metrics {
+		if basicFilter.Matches(metric) {
+			effectiveMetrics[metric] = true
+		}
+	}
+
+	return &metricsFilter{metadata, effectiveMetrics, basicFilter}, nil
+}
+
+// enabledMetrics returns list of metrics that are enabled via user-configuration or by
+// being included metrics.
+func (add *metricsFilter) enabledMetrics() []string {
+	metrics := make([]string, 0, len(add.extraMetrics)+len(add.metadata.IncludedMetrics))
+
+	for metric := range add.extraMetrics {
+		metrics = append(metrics, metric)
+	}
+
+	for metric := range add.metadata.IncludedMetrics {
+		metrics = append(metrics, metric)
+	}
+
+	return metrics
+}
+
+func (add *metricsFilter) shouldSend(dp *datapoint.Datapoint) bool {
+	if add.metadata.SendAll {
+		return true
+	}
+
+	if add.metadata.HasIncludedMetric(dp.Metric) {
+		// It's an included metric so send by default.
+		return true
+	}
+
+	if add.extraMetrics[dp.Metric] {
+		// User has explicitly chosen to send this metric (or a group that this metric belongs to).
+		return true
+	}
+
+	if add.metadata.MetricsExhaustive && !add.metadata.HasMetric(dp.Metric) {
```

Comment:
Not sure about these last two conditionals.

It seems like from this point down it should just return whether the metric matches the `add.stringFilter` and not worry about whether it is in the metadata list or not.

Reply:
Esp this conditional here would prevent somebody from being able to add a metric in if a developer made a typo in the metric name in the metadata.yaml.